pennylane/templates/subroutines/prepselprep.py

```diff
@@ -0,0 +1,144 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the PrepSelPrep template.
+This template contains a decomposition for performing a block-encoding on a
+linear combination of unitaries using the Prepare, Select, Prepare method.
+"""
+# pylint: disable=arguments-differ
+import copy
+
+import pennylane as qml
+from pennylane.operation import Operation
+
+
+class PrepSelPrep(Operation):
+    """This class implements a block-encoding of a linear combination of unitaries
+    using the Prepare, Select, Prepare method"""
+
+    def __init__(self, lcu, control, id=None):
+        coeffs, ops = lcu.terms()
+        control = qml.wires.Wires(control)
+        self.hyperparameters["lcu"] = lcu
+        self.hyperparameters["coeffs"] = coeffs
+        self.hyperparameters["ops"] = tuple(ops)
+        self.hyperparameters["control"] = control
+
+        if any(
+            control_wire in qml.wires.Wires.all_wires([op.wires for op in ops])
+            for control_wire in control
+        ):
+            raise ValueError("Control wires should be different from operation wires.")
+
+        target_wires = qml.wires.Wires.all_wires([op.wires for op in ops])
+        self.hyperparameters["target_wires"] = target_wires
+
+        all_wires = target_wires + control
+        super().__init__(*self.data, wires=all_wires, id=id)
+
+    def _flatten(self):
+        return tuple(self.lcu), (self.control)
+
+    @classmethod
+    def _unflatten(cls, data, metadata) -> "PrepSelPrep":
+        coeffs = [term.terms()[0][0] for term in data]
+        ops = [term.terms()[1][0] for term in data]
+        lcu = qml.ops.LinearCombination(coeffs, ops)
+        return cls(lcu, metadata)
+
+    def __repr__(self):
+        return f"PrepSelPrep(coeffs={tuple(self.coeffs)}, ops={self.ops}, control={self.control})"
+
+    def map_wires(self, wire_map: dict) -> "Select":
+        new_ops = [o.map_wires(wire_map) for o in self.hyperparameters["ops"]]
+        new_control = [wire_map.get(wire, wire) for wire in self.hyperparameters["control"]]
+        new_lcu = qml.dot(self.hyperparameters["coeffs"], new_ops)
+        return PrepSelPrep(new_lcu, new_control)
+
+    def decomposition(self):
+        return self.compute_decomposition(self.lcu, self.control)
+
+    @staticmethod
+    def compute_decomposition(lcu, control):
```

Comment:
It turns out that the `compute_decomposition` and `decomposition` functions are responsible for both decomposing the template (returning a list of the operations) and *queuing* the decomposition (See the note [here](https://docs.pennylane.ai/en/stable/code/api/pennylane.operation.Operator.html#pennylane.operation.Operator.compute_decomposition)). 

Typically, an operation automatically gets queued on creation, but we also have a function (qml.apply)[https://docs.pennylane.ai/en/stable/code/api/pennylane.apply.html?highlight=apply#pennylane.apply] to manually queue. 

It might also be helpful to see the implementation of the `compute_decomposition` method of `qml.QSVT` if you need another reference.

Reply:
As a follow up could you add a test to make sure we are queuing the decomposition?